src/System.Management.Automation/engine/PSConfiguration.cs

```diff
@@ -389,36 +389,42 @@ internal PSKeyword GetLogKeywords()
             if (!File.Exists(fileName)) { return defaultValue; }
 
             // Open file for reading, but allow multiple readers
-            fileLock.EnterReadLock();
-            try
+            if (fileLock.TryEnterReadLock(millisecondsTimeout: 10))
             {
-                // The config file can be locked by another process
-                // so we wait some milliseconds in 'WaitForFile()' for recovery before stop current process.
-                using (var readerStream = WaitForFile(fileName, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
-                using (var streamReader = new StreamReader(readerStream))
-                using (var jsonReader = new JsonTextReader(streamReader))
+                try
                 {
-                    var settings = new JsonSerializerSettings() { TypeNameHandling = TypeNameHandling.None, MaxDepth = 10 };
-                    var serializer = JsonSerializer.Create(settings);
-
-                    var configData = serializer.Deserialize<JObject>(jsonReader);
-                    if (configData != null && configData.TryGetValue(key, StringComparison.OrdinalIgnoreCase, out JToken jToken))
+                    // The config file can be locked by another process
+                    // so we wait some milliseconds in 'WaitForFile()' for recovery before stop current process.
+                    using (var readerStream = WaitForFile(fileName, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
+                    using (var streamReader = new StreamReader(readerStream))
+                    using (var jsonReader = new JsonTextReader(streamReader))
                     {
-                        return readImpl != null ? readImpl(jToken, serializer, defaultValue) : jToken.ToObject<T>(serializer);
+                        var settings = new JsonSerializerSettings() { TypeNameHandling = TypeNameHandling.None, MaxDepth = 10 };
+                        var serializer = JsonSerializer.Create(settings);
+
+                        var configData = serializer.Deserialize<JObject>(jsonReader);
+                        if (configData != null && configData.TryGetValue(key, StringComparison.OrdinalIgnoreCase, out JToken jToken))
+                        {
+                            return readImpl != null ? readImpl(jToken, serializer, defaultValue) : jToken.ToObject<T>(serializer);
+                        }
                     }
                 }
-            }
-            finally
-            {
-                fileLock.ExitReadLock();
+                catch (IOException)
+                {
+                    return defaultValue;
```

Comment:
As we discussed, might cause security settings to be bypassed.

Reply:
What is the security settings? I don't understand why it could cause a problem if defaults is secure too. If we want _force_ security we should use other means (group policy).

Reply:
All settings under `PowerShellPolicies` are security settings. An example is pasted below.
Currently, **powershell doesn't enforce them by default** when the config file doesn't exists or those settings are not defined in the config file.
If we want to stick to the default value solution for the race condition problem, additional work is needed to make it return the appropriate value when the file exists but not accessible.
But still, it's impossible to have an appropriate value for some of the settings in case the file is not accessible, for example, `ProtectedEventLogging`, you won't be able to know what certificate to use to encrypt the logging.

> If we want _force_ security we should use other means (group policy).

Linux/macOS doesn't have group policy.

```
{
  "Microsoft.PowerShell:ExecutionPolicy": "RemoteSigned",
  "PowerShellPolicies": {
    "ScriptExecution": {
      "ExecutionPolicy": "RemoteSigned",
      "EnableScripts": true
    },
    "ScriptBlockLogging": {
      "EnableScriptBlockInvocationLogging": true,
      "EnableScriptBlockLogging": true
    },
    "ModuleLogging": {
      "EnableModuleLogging": false,
      "ModuleNames": [
        "PSReadline",
        "PowerShellGet"
      ]
    },
    "ProtectedEventLogging": {
      "EnableProtectedEventLogging": false,
      "EncryptionCertificate": [
        "Joe"
      ]
    },
    "Transcription": {
      "EnableTranscripting": true,
      "EnableInvocationHeader": true,
      "OutputDirectory": "F:\\tmp\\new"
    },
    "UpdatableHelp": {
      "DefaultSourcePath": "f:\\temp"
    },
    "ConsoleSessionConfiguration": {
      "EnableConsoleSessionConfiguration": false,
      "ConsoleSessionConfigurationName": "name"
    }
  },
  "LogLevel": "verbose"
}
```

Reply:
> ProtectedEventLogging, you won't be able to know what certificate to use to encrypt the logging

Thanks! My question was just about such settings. I think we need a compromise. This is an open project and if someone needs a highly secure configuration, they can hardcode needed parameters in fork repo and safely ignore the config file at all. In the repo we can hardcode default parameters as needed too and allow PowerShell start without reading the config file. Also users can check current config in script/module/host application and terminate if they see wrong config.